jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -600,62 +569,101 @@ ecma_builtin_string_prototype_object_replace (ecma_value_t this_value, /**< this
+ecma_builtin_string_prototype_object_search (ecma_value_t this_value, /**< this argument */
+                                             ecma_value_t regexp_value) /**< routine's argument */
 {
-
-  ecma_value_t regexp_value = ECMA_VALUE_EMPTY;
-
-  if (ECMA_IS_VALUE_ERROR (ecma_builtin_string_prepare_search (regexp_arg, &regexp_value)))
+#if ENABLED (JERRY_ES2015)
+  if (!(ecma_is_value_undefined (regexp_value) || ecma_is_value_null (regexp_value)))
   {
-    return ECMA_VALUE_ERROR;
-  }
+    ecma_object_t *obj_p = ecma_get_object_from_value (ecma_op_to_object (regexp_value));
+    ecma_value_t search_symbol = ecma_op_object_get_by_symbol_id (obj_p, LIT_MAGIC_STRING_SEARCH);
+    ecma_deref_object (obj_p);
 
-  /* 5. */
-  ecma_string_t *this_string_p = ecma_get_string_from_value (to_string_value);
-  ecma_ref_ecma_string (this_string_p);
+    if (ECMA_IS_VALUE_ERROR (search_symbol))
+    {
+      return search_symbol;
+    }
 
-  ecma_value_t match_result = ecma_regexp_exec_helper (regexp_value, to_string_value, true);
+    if (!ecma_is_value_undefined (search_symbol) && !ecma_is_value_null (search_symbol))
+    {
+      if (!ecma_op_is_callable (search_symbol))
+      {
+        ecma_free_value (search_symbol);
+        return ecma_raise_type_error (ECMA_ERR_MSG ("@@search is not callable"));
+      }
+
+      ecma_object_t *search_method = ecma_get_object_from_value (search_symbol);
 
-  ecma_value_t ret_value = ECMA_VALUE_ERROR;
+      ecma_value_t arguments[] = { this_value };
+      ecma_value_t search_result = ecma_op_function_call (search_method, regexp_value, arguments, 1);
 
-  if (ECMA_IS_VALUE_ERROR (match_result))
+      ecma_deref_object (search_method);
+      return search_result;
+    }
+  }
+#else /* !ENABLED (JERRY_ES2015) */
+  if (ecma_object_is_regexp_object (regexp_value))
   {
-    goto cleanup;
+    return ecma_regexp_search_helper (regexp_value, this_value);
   }
+#endif /* ENABLED (JERRY_ES2015) */
 
-  ecma_number_t offset = -1;
+  ecma_value_t result = ECMA_VALUE_ERROR;
 
-  if (!ecma_is_value_null (match_result))
+  ecma_string_t *string_p = ecma_op_to_string (this_value);
+  if (string_p == NULL)
   {
-    JERRY_ASSERT (ecma_is_value_object (match_result));
-
-    ecma_object_t *match_object_p = ecma_get_object_from_value (match_result);
-
-    ecma_value_t index_value = ecma_op_object_get_by_magic_id (match_object_p, LIT_MAGIC_STRING_INDEX);
+    return result;
+  }
 
-    JERRY_ASSERT (!ECMA_IS_VALUE_ERROR (index_value) && ecma_is_value_number (index_value));
+  ecma_string_t *pattern_p = ecma_regexp_read_pattern_str_helper (regexp_value);
+  if (pattern_p == NULL)
+  {
+    goto cleanup_string;
+  }
 
-    offset = ecma_get_number_from_value (index_value);
+  ecma_value_t new_regexp = ecma_op_create_regexp_object (pattern_p, 0);
+  ecma_deref_ecma_string (pattern_p);
+  if (ECMA_IS_VALUE_ERROR (new_regexp))
+  {
+    goto cleanup_string;
```

Comment:
Shouldn't we need to set `result` to `new_regexp` (the error)?

Reply:
It's not necesary, `result` has been initialized to `ECMA_VALUE_ERROR` and has not been re-assigned since then.